mass-deployer/pkg/mass-deployer/deployer.go

```diff
@@ -0,0 +1,243 @@
+package deployer
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/sethvargo/go-retry"
+	"gopkg.in/yaml.v3"
+
+	"github.com/rs/zerolog/log"
+	"github.com/threefoldtech/tfgrid-sdk-go/grid-client/deployer"
+	"github.com/threefoldtech/tfgrid-sdk-go/grid-client/workloads"
+	"github.com/threefoldtech/zos/pkg/gridtypes"
+)
+
+func RunDeployer(cfg Config) error {
+	ctx := context.Background()
+	passedGroups := map[string][]string{}
+	failedGroups := map[string]error{}
+
+	tfPluginClient, err := setup(cfg)
+	if err != nil {
+		return fmt.Errorf("failed to create deployer: %v", err)
+	}
+
+	deploymentStart := time.Now()
+
+	for _, nodeGroup := range cfg.NodeGroups {
+		log.Info().Msgf("running deployer for node group %s", nodeGroup.Name)
+		firstTrial := true
+
+		if err := retry.Do(ctx, retry.WithMaxRetries(2, retry.NewConstant(1*time.Nanosecond)), func(ctx context.Context) error {
```

Comment:
we should maybe consider retrying only failed deployments not the whole group from the start. will save a lot of time I think